### ISSUE TITLE
Error "Enforcement literal not supported in constraint" (running v0.9.7 model)

### ISSUE THREAD

**Issue by hakank** [CPMpy/cpmpy#126]
The following 0.9.7 model (http://hakank.org/cpmpy/giant_cat_army_riddle.py) throws an `Enforcement literal not supported in constraint` error when running in 0.9.8: 
```
import sys
import numpy as np
from cpmpy import *
from cpmpy.solvers import *

def giant_cat_army_riddle(n=60):

  model = Model()

  maxval = 60

  x = intvar(0,maxval,shape=n,name="x")
  
  model = Model([AllDifferent(x)])

  # Given and symmetry breaking
  model += [x[0] == 0]

  # x[1] == 5 || x[1] == 7
  model += [(x[1] == 5) | (x[1] == 7)]
  model += (x[n-1] == 14)
  
  for i in range(n-1):
      model += [(x[i+1] == x[i] + 5) |
                (x[i+1] == x[i] + 7) |
                (x[i] == x[i+1]*x[i+1])]
      
  # 0 .. 2 .. 10 .. 14
  ix2 = intvar(1,n,name="ix2")
  ix10 = intvar(1,n,name="ix10")            
  model += [2 == x[ix2]]
  model += [10 == x[ix10]]            
  model += [ix2 < ix10]

  ss = CPM_ortools(model)

  if ss.solve():
    print("n:",n)
    xval = x.value()
    print("x:",xval)
    print("With the operations:")
    for i in range(n):
      if i > 0:
        if xval[i]-xval[i-1] == 5:
          print("(+5)", end=" ")
        elif xval[i]-xval[i-1] == 7:
          print("(+7)", end=" ")
        else: 
          print("(sqrt)", end=" ")
      print(xval[i],end=" ")
    print()
    print(ss.status())

    return True
  
  else:
    return None

# Try different sequence lengths
for n in range(5,60):
  sol = giant_cat_army_riddle(n)
  if sol != None:
    min_len = n 
    break
print("\nmin_len:",min_len)
```

The error:
```
Traceback (most recent call last):
  File "giant_cat_army_riddle.py", line 100, in <module>
    sol = giant_cat_army_riddle(n)
  File "giant_cat_army_riddle.py", line 75, in giant_cat_army_riddle
    if ss.solve():
  File "/usr/local/lib/python3.7/site-packages/cpmpy/solvers/ortools.py", line 170, in solve
    raise Exception("OR-Tools says: model invalid:", self.ort_model.Validate())
Exception: ('OR-Tools says: model invalid:', 'Enforcement literal not supported in constraint: enforcement_literal: 9 int_prod { vars: 1 vars: 1 }')
```

**Comment by hakank** [comment]
The same error is thrown for the following models:
- http://hakank.org/cpmpy/a_round_of_golf.py
- http://hakank.org/cpmpy/age_changing.py
- http://hakank.org/cpmpy/best_host.py
- http://hakank.org/cpmpy/exodus.py
- http://hakank.org/cpmpy/kenken2.py
- http://hakank.org/cpmpy/nadel.py
- http://hakank.org/cpmpy/number_puzzle8.py
- http://hakank.org/cpmpy/photo_problem.py
- http://hakank.org/cpmpy/rehearsal.py

**Comment by tias** [comment]
In response to the other bugreport and my suggestion of going from lazy to eager: it might not be sufficient, ortools only chokes (on this bug, and so perhaps also on the previous one) when calling solve.

So, for anybody potentially interested and for perhaps adding to the debugging guide, here is a another way to eagerly check which constraint it might be, after creating a model in fact:

```
  from cpmpy.transformations.flatten_model import flatten_constraint
  for c in model.constraints:
        print("Trying",c)
        print("Flat:",flatten_constraint(c))
        Model([c]).solve()
```

which results in the following smaller case:
```
from cpmpy.transformations.flatten_model import flatten_constraint
x,y = intvar(1,9, shape=2)
bv = boolvar()
c = (bv == (x == y*y))
print(c)
print(flatten_constraint(c))
Model(c).solve()
```

In other words: OR-tools does not support reifying a multiplication of two variables, but CPMpy tries it nonetheless.

It is quite surprising this worked in 0.9.7.
Then again, the or-tools translator was dirty-er then it is now, and it did some things right by chance, such as translating 'bv == (someconstraint == x)' to '(bv == (x == aux)) & (someconstraint == aux)'

That is the difference. In 'a round of golf' it is the same with the element constraint.

The solution is a better generic handling of reification of equalities. This could be seen as part of #105 but it is also not.

Hmm... hmm... needs a bit more thought/care, but urgent to fix indeed.

**Comment by hakank** [comment]
Thanks for the great debugging tip, Tias.

For now, I'll simply change to the MiniZinc solver.

**Comment by tias** [comment]
We implemented the reification handling in #127 that is now merged, so now these should all be properly supported by ortools

**Comment by hakank** [comment]
I can confirm that all the models now work with CPM_ortools. Great!
